[WIP]: Remote Cluster

Shrinjay opened the pull request.

Shrinjay committed:
add remote cluster functionality

Shrinjay committed:
Add context selection to helm and passthrough

Shrinjay committed:
Remove label

Shrinjay committed:
Put back comments

Shrinjay committed:
Fix names

DerekTBrown commented on enterprise_gateway/services/external/k8s_client_factory.py:
```diff
@@ -0,0 +1,31 @@
+        if os.getenv("KUBERNETES_SERVICE_HOST"):
+            # Running inside cluster
+            if os.getenv('EG_USE_REMOTE_CLUSTER') and get_remote_if_available:
+                kubeconfig_path = os.environ.get('EG_REMOTE_CLUSTER_KUBECONFIG_PATH', '/etc/kube/config')
+                context = os.environ.get('EG_REMOTE_CLUSTER_CONTEXT', None)
```
Nit: I would be consistent with the method used for env fetching. I don't think there is a big difference, but I think it is wise to pick one or another.
```suggestion
                kubeconfig_path = os.getenv('EG_REMOTE_CLUSTER_KUBECONFIG_PATH', '/etc/kube/config')
                context = os.getenv('EG_REMOTE_CLUSTER_CONTEXT', None)
```

DerekTBrown commented on enterprise_gateway/services/external/k8s_client_factory.py:
```diff
@@ -0,0 +1,31 @@
+        if os.getenv("KUBERNETES_SERVICE_HOST"):
+            # Running inside cluster
+            if os.getenv('EG_USE_REMOTE_CLUSTER') and get_remote_if_available:
+                kubeconfig_path = os.environ.get('EG_REMOTE_CLUSTER_KUBECONFIG_PATH', '/etc/kube/config')
```
Did you pick the `/etc/kube/config` default, or did this come from something existing?

Shrinjay replied:
I picked that, honestly by random, I realize I forgot to set this as the default in the helm chart as well, but the idea was for the path to be the same in the helm chart and in code

DerekTBrown commented on enterprise_gateway/services/external/k8s_client_factory.py:
```diff
@@ -0,0 +1,31 @@
+        super().__init__()
+
+    def get_kubernetes_client(self, get_remote_if_available=True) -> client.ApiClient:
+        """Get kubernetes api client with appropriate configuration"""
```
Can you document what this does and also what the effect of `get_remote_if_available` is?

DerekTBrown commented on enterprise_gateway/services/external/k8s_client_factory.py:
```diff
@@ -0,0 +1,31 @@
+            config.load_kube_config(client_configuration=kubernetes_config)
+
+        self.log.debug(
+            "Created kubernetes client for host {host}".format(host=kubernetes_config.host)
```
Nit: `f-strings` for the win
```suggestion
            f"Created kubernetes client for host {kubernetes_config.host}"
```

DerekTBrown commented on enterprise_gateway/services/processproxies/k8s.py:
```diff
@@ -308,14 +312,61 @@ def _create_kernel_namespace(self, service_account_name: str) -> str:
         return namespace
 
+    def _create_service_account_if_not_exists(self, namespace: str, service_account_name: str) -> None:
+        service_account_list_in_namespace: client.V1ServiceAccountList = client\
+            .CoreV1Api(api_client=kubernetes_client)\
+            .list_namespaced_service_account(namespace=namespace)
```
(More of a meta point than a specific set of lines)

Slash continuations in Python are pretty evil.  Was this something you put in manually, or something the formatter did for you?

You might want to put these on one (really long) line, and then use `black` to find a solution that doesn't involve slashes. I believe the project uses `black` anyway, so it is a good idea to run that over your code from the getgo.

Shrinjay replied:
PyCharm did it when I tried to chain the invocations over multiple lines like I usually do in js/ts, honestly wasn't aware slash continuations were a problem until I just googled it, so thanks for pointing this one out

DerekTBrown commented on enterprise_gateway/services/processproxies/k8s.py:
```diff
@@ -308,14 +312,61 @@ def _create_kernel_namespace(self, service_account_name: str) -> str:
+            .list_namespaced_service_account(namespace=namespace)
+
+        service_accounts_in_namespace: List[client.V1ServiceAccount] = service_account_list_in_namespace.items
+        service_account_names_in_namespace: List[str] = list(map(lambda svcaccount: svcaccount.metadata.name, service_accounts_in_namespace))
+
+        if service_account_name not in service_account_names_in_namespace:
```
1. Super super good on you for using functional programming here. This is a senior dev move.

2. We can make this an even more big-headed by using a [generator comprehension](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Generators_and_Comprehensions.html#Creating-your-own-generator:-generator-comprehensions). The list/map method is on the order of n in terms of memory complexity and roughly 2n in terms of time complexity (we build a list of length n which takes n time, then we scan the list which takes n time). The lambda also requires constructing a function context, which can be expensive. The generator method is on the order of o(1) in terms of memory complexity and takes only n in terms of time complexity (we just perform a single scan of the other list without actually constructing anything). 

```suggestion
        service_account_names_in_namespace = (svcaccount.metadata.name for svcaccount in service_accounts_in_namespace)
        
        if service_account_name not in service_account_names_in_namespace:
```

DerekTBrown replied:
At this point, you are probably thinking "why do we care about performance, this is not that performance intensive?" 

This has less to do with performance and more to do with code quality; we want the Juypter reviewers to think "damn, thats some slick code, they must know what they are doing."

Shrinjay replied:
Oh I love finding out about optimizations like this, I honestly don't know enough about Python for how much I use it, so thanks for teaching me about this 😄

Shrinjay replied:
Oh it's list comprehension... darn I should've remembered about that, I keep defaulting to the list-map combo, probably because I use .map a lot in TS

DerekTBrown commented on enterprise_gateway/services/processproxies/k8s.py:
```diff
@@ -308,14 +312,61 @@ def _create_kernel_namespace(self, service_account_name: str) -> str:
+        remote_cluster_roles: client.V1ClusterRoleList = client\
+            .RbacAuthorizationV1Api(api_client=kubernetes_client)\
+            .list_cluster_role()
+        remote_cluster_role_names = list(map(lambda role: role.metadata.name, remote_cluster_roles.items))
+
+        if kernel_cluster_role not in remote_cluster_role_names:
```
Same as above RE: generator comprehension

Shrinjay committed:
Fix docstrings, generator, and various other

Shrinjay committed:
black fixes

DerekTBrown commented on enterprise_gateway/services/processproxies/k8s.py:
```diff
@@ -308,14 +312,61 @@ def _create_kernel_namespace(self, service_account_name: str) -> str:
+
+            self.log.info(f"Created service account {service_account_name} in namespace {namespace}")
+
+    def _forward_role_to_remote(self) -> None:
```
A few thoughts:

1. I don't think we (Lacework) want to grant a ClusterRole to JEG. At most, I think we want to grant roles to a handful of namespaces. 
2. I don't think we should be copying a role from one place to another, this seems non-transparent.
3. I would recommend adding a static rolebinding to a specific namespace.
3. I would recommend adding an option to _not_ create the service account, allowing the user to supply their own service account. This lets people customize as they choose.

DerekTBrown commented on enterprise_gateway/services/external/k8s_client_factory.py:
```diff
@@ -0,0 +1,31 @@
+        kubernetes_config: client.Configuration = client.Configuration()
+        if os.getenv("KUBERNETES_SERVICE_HOST"):
+            # Running inside cluster
+            if os.getenv('EG_USE_REMOTE_CLUSTER') and get_remote_if_available:
```
I am not sure this works as expected. This will evaluate to true if `EG_USE_REMOTE_CLUSTER` is set to _anything_ (including `EG_USE_REMOTE_CLUSTER=False`). 

We should confirm this has the same semantics as other env options defined in JEG. Do they look for True/False or just Set/Unset. If True/False, we should add a more rigorous check. If set/unset, we should use `EG_USE_REMOTE_CLUSTER=1` in the helm chart to denote this.

Shrinjay commented:
The reason I added the copying was more as a first pass than anything. The
issue is we want the role created in the remote cluster to be configurable
via Helm. The only way to inject configurations from helm into runtime is
environment variables. I’m not quite sure if we want to configure the role
for the remote using environment variables, hence the copying. The role in
the local space is defined by the helm chart so we just copy that over.

On Wed, Jan 25, 2023 at 4:23 PM Derek Brown ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In enterprise_gateway/services/processproxies/k8s.py
> <https://github.com/Shrinjay/enterprise_gateway/pull/1#discussion_r1087281978>
> :
>
> > +        if service_account_name not in service_account_names_in_namespace:
> +            service_account_metadata = {
> +                "name": service_account_name
> +            }
> +            service_account_to_create: client.V1ServiceAccount = client.V1ServiceAccount(
> +                kind="ServiceAccount",
> +                metadata=service_account_metadata
> +            )
> +
> +            client\
> +                .CoreV1Api(api_client=kubernetes_client)\
> +                .create_namespaced_service_account(namespace=namespace, body=service_account_to_create)
> +
> +            self.log.info(f"Created service account {service_account_name} in namespace {namespace}")
> +
> +    def _forward_role_to_remote(self) -> None:
>
> A few thoughts:
>
>    1. I don't think we (Lacework) want to grant a ClusterRole to JEG. At
>    most, I think we want to grant roles to a handful of namespaces.
>    2. I don't think we should be copying a role from one place to
>    another, this seems non-transparent.
>    3. I would recommend adding a static rolebinding to a specific
>    namespace.
>    4. I would recommend adding an option to *not* create the service
>    account, allowing the user to supply their own service account. This lets
>    people customize as they choose.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Shrinjay/enterprise_gateway/pull/1#pullrequestreview-1270268984>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO5NQAWJ6YWC2OZ6RNLSOOTWUGYXTANCNFSM6AAAAAAUFVOHUI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Shrinjay commented:
I realized I'm an idiot and configmaps exist... we can mount the local role resource definitions as a configmap and then create it in the remote rather than copying the role.

Shrinjay committed:
Add remote autoconfiguration and role

Shrinjay committed:
Fix env var checking

Shrinjay committed:
move to namespaced roles for remote clusters

Shrinjay committed:
Formatting and comments

Shrinjay committed:
Add documentation and cleanup

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Add more linting (#1245)

* add more linting
* Address codeQL alert

---------

Co-authored-by: Kevin Bates <kbates4@gmail.com>

Shrinjay committed:
Fix doc requirements

Shrinjay committed:
Merge branch 'feature/remote-cluster' of https://github.com/Shrinjay/enterprise_gateway into feature/remote-cluster

Shrinjay committed:
Merge branch 'main' into feature/remote-cluster

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Shrinjay committed:
Change to external cluster naming

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Shrinjay committed:
Chart lint fixes

Shrinjay committed:
Merge branch 'feature/remote-cluster' of https://github.com/Shrinjay/enterprise_gateway into feature/remote-cluster

lresende committed:
Update java certs to avoid multi-arch builds to hang (#1241)

* Update java certs to avoid multi-arch builds to hang
* Update additional images use by spark kernels
* Fix kernel-spark-r dockerfile
* Remove sudo from dpkg command
* Add cleanup after install
* Update dockerfile to install ca-certificates-java
* install java certs only after headless jre

---------

Co-authored-by: Kevin Bates <kbates4@gmail.com>

lresende committed:
Dump Kubernetes YAML/CRD on kernel startup fails (#1249)

lresende committed:
Change demo-base container build from conda to mamba (#1250)

* Change demo-base container build from conda to mamba
* Continue using continuumio/miniconda3 image but with mamba binaries

dependabot[bot] committed:
Bump black[jupyter] from 22.12.0 to 23.1.0 (#1251)

* Bump black[jupyter] from 22.12.0 to 23.1.0

Bumps [black[jupyter]](https://github.com/psf/black) from 22.12.0 to 23.1.0.
- [Release notes](https://github.com/psf/black/releases)
- [Changelog](https://github.com/psf/black/blob/main/CHANGES.md)
- [Commits](https://github.com/psf/black/compare/22.12.0...23.1.0)

---
updated-dependencies:
- dependency-name: black[jupyter]
  dependency-type: direct:production
  update-type: version-update:semver-major
...

* Apply black formatting
---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

dependabot[bot] committed:
Bump ruff from 0.0.237 to 0.0.241 (#1252)

* Bump ruff from 0.0.237 to 0.0.241

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.237 to 0.0.241.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.237...v0.0.241)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...
* Add config for max-args and max-statements
* Lower max-statements, update ruff version in pre-commit

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#1253)

* [pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/python-jsonschema/check-jsonschema: 0.19.2 → 0.21.0](https://github.com/python-jsonschema/check-jsonschema/compare/0.19.2...0.21.0)
- [github.com/psf/black: 22.12.0 → 23.1.0](https://github.com/psf/black/compare/22.12.0...23.1.0)
- [github.com/charliermarsh/ruff-pre-commit: v0.0.237 → v0.0.242](https://github.com/charliermarsh/ruff-pre-commit/compare/v0.0.237...v0.0.242)

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump ruff from 0.0.241 to 0.0.246 (#1254)

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.241 to 0.0.246.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.241...v0.0.246)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

starskyreverie committed:
A few small changes to gateway client (#1257)

* Update gateway_client.py
* Fix linting :)
* Fix execute calls everywhere
* fix tests
* Fix log
* try except websocket connection

dependabot[bot] committed:
Bump ruff from 0.0.246 to 0.0.249 (#1256)

* Bump ruff from 0.0.246 to 0.0.249

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.246 to 0.0.249.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.246...v0.0.249)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

* Fix lint errors
* Selectively ignore C901 - function too complex

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

kevin-bates committed:
Fix badges on README (#1259)

* Fix README badges
* Remove Gitter badge due to its migration to new provider
* Remove trailing vertical divide on header

manoj-jeswani committed:
Fix the flow of kernel-class-name option in bootstrap script (#1260)

Co-authored-by: Manoj Jeswani <m_jeswani@apple.com>

dependabot[bot] committed:
Bump ruff from 0.0.249 to 0.0.253 (#1264)

* Bump ruff from 0.0.249 to 0.0.253

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.249 to 0.0.253.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.249...v0.0.253)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Ignore PLW0603 when linting

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

Commit:
decouple image name fetching from image puller (#1263)

* decouple image name fetching from the image puller to allow dynamic image name fetching

* fix image_list not initialize issue

* fix typo

* remove noqa

* refactoring

* Address lint failures

* add docs

* update k8s docs

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

kevin-bates committed:
Set language and fix conda env link (#1265)

lresende committed:
Handle not found exception during remove_kernel (#1262)

lresende committed:
Bump to 3.2.1

lresende committed:
Prepare release 3.2.1

lresende committed:
Bump to 3.3.0.dev0

lresende committed:
Prepare for next development interaction 3.3.0.dev0

kevin-bates committed:
Reset SPARK_VERSION to 3.2.1 (#1268)

starskyreverie committed:
Replace "is defined" in Jinja templates if conditions (#1267)

* Update gateway_client.py
* remove trailing whitespace
* Fix execute calls everywhere
* fix tests
* Fix log
* Remove 'is defined' from jinja templates
* Remove remaining 'is defined's
* Add  back to numeric variables in Jinja templates
* Add  back to kernel uid/gid

lresende committed:
Update test-install task to build dist as dep (#1270)

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#1273)

updates:
- [github.com/charliermarsh/ruff-pre-commit: v0.0.253 → v0.0.254](https://github.com/charliermarsh/ruff-pre-commit/compare/v0.0.253...v0.0.254)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

lresende committed:
Set restart strategy to never for spark operator (#1269)

* Set restart strategy to never for spark operator
* Disable eviction for driver pod
* Disable eviction on Kubernetes python kernels as well

dependabot[bot] committed:
Bump ruff from 0.0.253 to 0.0.254 (#1272)

* Bump ruff from 0.0.253 to 0.0.254

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.253 to 0.0.254.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.253...v0.0.254)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update ruff version in pre-commit config

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

lresende committed:
Properly propagate errors on Spark Operator CRD failures (#1271)

* Properly report failures on Spark Operator kernels
* Update helper methods to get initial and error states
* Try to use detect_launch_failure as an option to identify failed CRD submission
* Fix small typos on lower conversion
* Update logic to better handle kernel connection parameters
* Fix restart issues
* Remove detect_launch_failure unnecessary code
* Remove debug code
* Simplify get_container_status signature
* Remove lower after the signature change
* Move get_initial_states to CustomResourceProcessProxy
* Conditionally display get_container_status debug message
* Move get_container_status to CRD process proxy
* Update get_initial_states function doc to emphasize lowercase contract
* Reword to remove spark operator references
* Add log updates
* Remove obsolete code
* Add more fields on the log statement
* Update log logic
* More logging fix-ups

---------

Co-authored-by: Kevin Bates <kbates4@gmail.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

Commit:
Fix docker swarm not being to start due to missing inherited function (#1277)

Chiragasourabh committed:
Updating base image tag to latest available tag (#1280)

starskyreverie committed:
Filter sensitive values from being logged in ProcessProxy (#1279)

dependabot[bot] committed:
Bump ruff from 0.0.254 to 0.0.255 (#1278)

* Bump ruff from 0.0.254 to 0.0.255

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.254 to 0.0.255.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.254...v0.0.255)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Additional changes necessary for ruff-0.0.255

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

starskyreverie committed:
Move shutdown kernel code to be within KernelClient (#1281)

* GatewayClient changes

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Initialize variables before websocket.create_connection in KernelClient.__init__()

* Refactor fail on WS connection

* Avoid need for attribute checks

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

lresende committed:
Bump to 3.2.2

lresende committed:
Prepare release 3.2.2

lresende committed:
Bump to 3.3.0.dev0

lresende committed:
Prepare for next development interaction 3.3.0.dev0

dependabot[bot] committed:
Bump ruff from 0.0.255 to 0.0.257 (#1282)

* Bump ruff from 0.0.255 to 0.0.257

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.255 to 0.0.257.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.255...v0.0.257)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Address changes for ruff upgrade

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

dependabot[bot] committed:
Bump ruff from 0.0.257 to 0.0.259 (#1283)

* Bump ruff from 0.0.257 to 0.0.259

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.257 to 0.0.259.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.257...v0.0.259)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Ignore S311 - only used for port selection within range and suffix

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

dependabot[bot] committed:
Bump ruff from 0.0.259 to 0.0.260 (#1285)

* Bump ruff from 0.0.259 to 0.0.260

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.259 to 0.0.260.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.259...v0.0.260)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update ruff version in pre-commit config

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

dependabot[bot] committed:
Bump black[jupyter] from 23.1.0 to 23.3.0 (#1284)

* Bump black[jupyter] from 23.1.0 to 23.3.0

Bumps [black[jupyter]](https://github.com/psf/black) from 23.1.0 to 23.3.0.
- [Release notes](https://github.com/psf/black/releases)
- [Changelog](https://github.com/psf/black/blob/main/CHANGES.md)
- [Commits](https://github.com/psf/black/compare/23.1.0...23.3.0)

---
updated-dependencies:
- dependency-name: black[jupyter]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update black version in pre-commit config

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#1286)

updates:
- [github.com/python-jsonschema/check-jsonschema: 0.21.0 → 0.22.0](https://github.com/python-jsonschema/check-jsonschema/compare/0.21.0...0.22.0)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump ruff from 0.0.260 to 0.0.261 (#1288)

* Bump ruff from 0.0.260 to 0.0.261

Bumps [ruff](https://github.com/charliermarsh/ruff) from 0.0.260 to 0.0.261.
- [Release notes](https://github.com/charliermarsh/ruff/releases)
- [Changelog](https://github.com/charliermarsh/ruff/blob/main/BREAKING_CHANGES.md)
- [Commits](https://github.com/charliermarsh/ruff/compare/v0.0.260...v0.0.261)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update ruff version in pre-commit config

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kevin Bates <kbates4@gmail.com>

Shrinjay committed:
Update docs/source/operators/deploy-kubernetes.md

Co-authored-by: Kevin Bates <kbates4@gmail.com>

Shrinjay committed:
Update docs/source/operators/deploy-kubernetes.md

Co-authored-by: Kevin Bates <kbates4@gmail.com>

Shrinjay committed:
Update enterprise_gateway/services/processproxies/crd.py

Co-authored-by: Kevin Bates <kbates4@gmail.com>

Shrinjay committed:
Apply suggestions from code review

Co-authored-by: Kevin Bates <kbates4@gmail.com>

Shrinjay committed:
Update docs/source/operators/deploy-kubernetes.md

Co-authored-by: Kevin Bates <kbates4@gmail.com>

Shrinjay committed:
Update etc/kubernetes/helm/enterprise-gateway/templates/deployment.yaml

Co-authored-by: Kevin Bates <kbates4@gmail.com>

Shrinjay committed:
progress

Shrinjay committed:
throw warning

Shrinjay committed:
merge latest

Shrinjay committed:
Merge branch 'main' into feature/remote-cluster

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
